### PR TITLE
new type-erased memory resources

### PR DESCRIPTION
**Pull request by ericniebler**
## Description

this PR reimplements `cudax::mr::any_resource` and adds a new `cudax::mr::resource_ref`. they both pass their existing regressions tests without modification. the new types also provide a `try_get_property` function that can retrieve a property that was "sliced off" during an interface-narrowing conversion.

neither `cudax::mr::any_resource` nor `cudax::mr::resource_ref` need to store the vtable entries for the properties _in-situ_, which results in a space savings proportional to the number of properties.

there is extra code to make `cudax::mr::any_resource` efficiently convertible to the existing `cuda::mr::resource_ref`.

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  template <class...>
+  struct __iproperty : interface<__iproperty>
+  {
+    _CUDAX_API friend auto get_property(const __iproperty& __obj, _Property __prop) noexcept //
+      -> __property_result_t<_Property>
+    {
+      if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+      {
+        return __cudax::virtcall<&__get_property<__iproperty>>(&__obj, __prop);
+      }
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:81]: Is that really necessary? It might actually give grief with compilers that mess up `if constexpr`
```suggestion
      return __cudax::virtcall<&__get_property<__iproperty>>(&__obj, __prop);
```

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+{
+  template <class _Ty>
+  _CUDAX_PUBLIC_API static auto __get_property(const _Ty& __obj, _Property __prop) noexcept //
+    -> __property_result_t<_Property>
+  {
+    static_assert(noexcept(get_property(__obj, __prop)));
+    if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return get_property(__obj, __prop);
+    }
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:69]: Ditto we should just be able to directly return
```suggestion
      return get_property(__obj, __prop);
```

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+};
+
+template <class _Property>
+using __iproperty = typename __with_property<_Property>::template __iproperty<>;
+
+template <class... _Properties>
+using __iproperty_set = iset<__iproperty<_Properties>...>;
+
+template <class...>
+struct __ibasic_resource : interface<__ibasic_resource, extends<icopyable<>, iequality_comparable<>>>
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:96]: Does this work if a resource is immovable?

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:96]: no resources need to be copyable

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh (hunk 1 above)
Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:81]: which compilers mess up `if constexpr`? we really don't want to make an indirect function call if we know statically that it's a no-op.

Comment (miscco) [reply cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:81]: I believe NVCC before 12.0 will most likely complain about missing return statement because it has faulty if constexpr handling. You will need at least a `_CCCL_UNREACHABLE()` at the end

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:80]: but it isn't unreachable. when the `if constexpr` branch isn't taken, the function returns `void`, so falling off the end if the right thing to do.

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  {
+    static_assert(noexcept(get_property(__obj, __prop)));
+    if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return get_property(__obj, __prop);
+    }
+    else
+    {
+      return void();
+    }
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:72]: This should rather be 
```suggestion
  _CCCL_UNREACHABLE();
```
Applies throughout

File: cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    -> __property_result_t<_Property>
+  {
+    static_assert(noexcept(get_property(__obj, __prop)));
+    if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return get_property(__obj, __prop);
+    }
+    else
+    {
+      return void();
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:71]: Should this just be
```suggestion
      return;
```

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh:71]: sadly not, for nvc++

File: cudax/test/memory_resource/any_resource.cu [hunk 6]
```diff
@@ -8,27 +8,39 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include <cuda/experimental/memory_resource.cuh>
+#ifndef __CUDA_ARCH__
```
Comment (pciolkosz) [inline cudax/test/memory_resource/any_resource.cu:11]: Why is this needed?

**Review by pciolkosz — APPROVED** [review]
The memory resource part looks good.
I will try to review the basic_any part in its PR too

File: cudax/include/cuda/experimental/__memory_resource/any_resource.cuh [hunk 7]
```diff
@@ -41,291 +41,719 @@
...
+// exceeds the lifetime of the memory resource used to allocate the storage
+template <class... _Properties>
+struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES any_async_resource
+    : basic_any<__iasync_resource<_Properties...>>
+    , __with_try_get_property<any_async_resource<_Properties...>>
+{
+  using any_async_resource::basic_any::basic_any;
+
+private:
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
```
Comment (pciolkosz) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:299]: I was wondering since the sliced off properties are now accessible with the `try_get_property` if we could drop this requirement?

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:299]: i lean towards "no". @miscco?

File: cudax/test/memory_resource/any_resource.cu (hunk 6 above)
Comment (ericniebler) [reply cudax/test/memory_resource/any_resource.cu:11]: the type-erased memory resources are host-only, no?

File: cudax/include/cuda/experimental/__memory_resource/any_resource.cuh [hunk 8]
```diff
@@ -41,310 +41,825 @@
...
-  static constexpr bool __valid_resource =
-    _Alloc_type == _CUDA_VMR::_AllocType::_Async
-      ? _CUDA_VMR::async_resource_with<_Resource, _Properties...>
-      : _CUDA_VMR::resource_with<_Resource, _Properties...>;
+template <class _Property>
+using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::__type_call1< //
+  _CUDA_VSTD::conditional_t<cuda::property_with_value<_Property>,
+                            _CUDA_VSTD::__type_quote1<__property_value_t>,
+                            _CUDA_VSTD::__type_always<void>>,
+  _Property>;
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:61]: I find this overly complicated compared to just
```suggestion
using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::conditional_t<cuda::property_with_value<_Property>,__property_value_t<_Property>, void>;
```

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:61]: when `property_with_value<_Property>` is `false`, `__property_value_t<_Property>` is ill-formed.

**Review by miscco — APPROVED** [review]

File: cudax/include/cuda/experimental/__memory_resource/any_resource.cuh [hunk 9]
```diff
@@ -41,310 +41,847 @@
...
+//! @sa any_async_resource
+//! @sa resource_ref
+//! @sa async_resource_ref
+template <_ResourceKind _Kind, class... _Properties>
+class basic_any_resource
+{
+public:
+  //! @brief Constructs a \c basic_any_resource with no value
+  //! @post `has_value()` is `false`
+  basic_any_resource() noexcept;
```
Comment (miscco) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:455]: Maybe I am stupid, where is the implementation?

Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:455]: this whole block, from line 417 to line 861, is strictly for Doxygen's benefit. the actually implementation of, eg, `any_resource` is the following from line 301:

```c++
template <class... _Properties>
struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES any_resource
    : basic_any<__iresource<_Properties...>>
    , __with_try_get_property<any_resource<_Properties...>>
{
  using any_resource::basic_any::basic_any;

  // any_async_resource is convertible to any_resource
  _CCCL_TEMPLATE(class... _OtherProperties)
  _CCCL_REQUIRES(_CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__type_set<_OtherProperties...>, _Properties...>)
  any_resource(any_async_resource<_OtherProperties...> __other) noexcept
      : basic_any<__iresource<_Properties...>>(static_cast<basic_any<__iasync_resource<_OtherProperties...>>&>(__other))
  {}

private:
  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
                "The properties of cuda::experimental::any_resource must contain at least one execution space "
                "property!");
  using any_resource::basic_any::interface;
};
```

I felt that Doxygen comments for this class template would not be very enlightening, so instead i mocked up the user-facing interface and added doc comments to that.

File: cudax/include/cuda/experimental/__memory_resource/any_resource.cuh [hunk 10]
```diff
@@ -41,310 +41,847 @@
...
-  //! @brief Inequality comparison between two \c basic_any_resource
-  //! @param __rhs The other \c basic_any_resource
-  //! @return Checks whether both resources have the same equality function stored in their vtable and if so returns
-  //! the inverse result of that equality comparison. Otherwise returns true.
-  _CCCL_TEMPLATE(class... _OtherProperties)
-  _CCCL_REQUIRES((sizeof...(_Properties) == sizeof...(_OtherProperties))
-                   _CCCL_AND __properties_match<_OtherProperties...>)
-  _CCCL_NODISCARD bool operator!=(const basic_any_resource<_Alloc_type, _OtherProperties...>& __rhs) const
+_CCCL_DIAG_PUSH
+_CCCL_DIAG_SUPPRESS_GCC("-Wunused-but-set-parameter")
```
[the remark this answers is not in the document]
Comment (ericniebler) [reply cudax/include/cuda/experimental/__memory_resource/any_resource.cuh:199]: i know it's nutty. there never was an unused argument. older gcc warns anyway.

**Comment by miscco** [comment]
🥳
